src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp

```diff
@@ -269,116 +397,354 @@ void DefaultCoordinator::finalizeReadingState()
 
-        response.description.push_back({
-            .info = part->description.info,
-            .ranges = {},
-        });
+        RangesInDataPartDescription result{.info = part.description.info};
 
-        while (!part->description.ranges.empty() && current_mark_size < min_number_of_marks)
+        while (!part.description.ranges.empty() && current_marks_amount < min_number_of_marks)
         {
-            auto & range = part->description.ranges.front();
-            const size_t needed = min_number_of_marks - current_mark_size;
+            auto & range = part.description.ranges.front();
 
-            if (range.getNumberOfMarks() > needed)
+            /// Parts are divided into segments of `mark_segment_size` granules staring from 0-th granule
+            for (size_t segment_begin = roundDownToMultiple(range.begin, mark_segment_size);
+                 segment_begin < range.end && current_marks_amount < min_number_of_marks;
+                 segment_begin += mark_segment_size)
             {
-                auto range_we_take = MarkRange{range.begin, range.begin + needed};
-                response.description.back().ranges.emplace_back(range_we_take);
-                current_mark_size += range_we_take.getNumberOfMarks();
+                const auto cur_segment
+                    = MarkRange{std::max(range.begin, segment_begin), std::min(range.end, segment_begin + mark_segment_size)};
 
-                range.begin += needed;
-                break;
+                const auto owner = computeConsistentHash(part.description.info.getPartNameV1(), segment_begin, scan_mode);
+                if (owner == replica_num)
+                {
+                    const auto taken = takeFromRange(cur_segment, min_number_of_marks, current_marks_amount, result);
+                    if (taken == range.getNumberOfMarks())
+                        part.description.ranges.pop_front();
+                    else
+                    {
+                        range.begin += taken;
+                        break;
+                    }
+                }
+                else
+                {
+                    chassert(scan_mode == ScanMode::TakeWhatsMineByHash);
+                    enqueueSegment(part.description.info, cur_segment, owner);
+                    range.begin += cur_segment.getNumberOfMarks();
+                    if (range.getNumberOfMarks() == 0)
+                        part.description.ranges.pop_front();
+                }
             }
-
-            response.description.back().ranges.emplace_back(range);
-            current_mark_size += range.getNumberOfMarks();
-            part->description.ranges.pop_front();
         }
+
+        if (!result.ranges.empty())
+            response.description.push_back(std::move(result));
+    }
+}
+
+void DefaultCoordinator::selectPartsAndRanges(
+    size_t replica_num, ScanMode scan_mode, size_t min_number_of_marks, size_t & current_marks_amount, ParallelReadResponse & response)
```

Comment:
Let's separate protocol and implementation. There is no need to pass the response as a parameter here, only `response.description` (`RangesInDataPartsDescription`) is used through the calls.